lib/services/api/href_parser.rb

```diff
@@ -0,0 +1,72 @@
+module Api
+  class HrefParser
+    def self.parse(href)
+      new(href).parse
+    end
+
+    def initialize(href)
+      @href = href
+    end
+
+    def parse
+      return [nil, nil] unless href
+      if subcollection?
+        [subcollection.to_sym, ApplicationRecord.uncompress_id(subcollection_id)]
+      else
+        [collection.to_sym, ApplicationRecord.uncompress_id(collection_id)]
+      end
+    end
+
+    private
+
+    attr_reader :href
+
+    def subcollection?
+      !!subcollection
+    end
+
+    def path
```

Comment:
This could use some kind of caching, and how you go about it depends on how you decide to go forward with my previous comment I guess.

But for example, if you had a `href` that was a `subcollection`, and `#parse` was called this method would get executed 3 times and return the same value each time, but require doing the transformations and regexp evaluations each time:

* First time would be in line `13` when `subcollection?` is called (from `#subcollection => #path_parts => #path`)
* Second would be `subcollection.to_sym` in line 14
* Third would be from `subcollection_id` in line 14 as well

Oh!  And I lied, it would actually call this method 6 times because for each of the above, the call to `version?` also hits `path` (again, through `#path_parts` in that method).  Either caching the value of `#path_parts` via an instance variable, or storing the value of `path_parts` as something like `parts` (if you go with the "class method approach" from above) would avoid this duplicated work.

Reply:
I am happy to memoize if keeping the instance implementation is agreeable?

Reply:
Yup, using an instance is fine (elaborated in my response above).  👍